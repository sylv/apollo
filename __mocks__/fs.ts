import { fs as mockFs } from "memfs";
import { ufs } from "unionfs";

const realFs = jest.requireActual("fs");
ufs.use(mockFs as any).use(realFs);

export default ufs;
module.exports = ufs;
