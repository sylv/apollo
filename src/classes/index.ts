export * from "./Apollo";
export * from "./ApolloParser";
