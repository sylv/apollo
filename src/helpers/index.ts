export * from "./clean-file-path";
export * from "./clean-raw-title";
export * from "./get-all-matches";
export * from "./strip-space-placeholders";
