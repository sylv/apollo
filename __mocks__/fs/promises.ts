import fs from "../fs";

module.exports = fs.promises;
export default fs.promises;
