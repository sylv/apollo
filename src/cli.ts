import meow = require('meow');
import path = require('path');
import fs = require('fs');
import { Apollo } from './';
import { Logger } from './logger';
import { pathExists } from './helpers/pathExists';

const cli = meow(
  `
    Usage
      parcel-cli <input-directory> <output-directory>

    Example
      parcel-cli ./torrents ./clean
      parcel-cli ./torrents/**/*.mkv ./clean
      parcel-cli D:\\Torrents E:\\Media

    Options
      --debug, -D Enable development logging
      --min-size, --minSize The minimum size of a media file in bytes to be considered valid. Defaults to 25,000,000 (25Mb)
`,
  {
    flags: {
      debug: {
        type: 'boolean',
        alias: ['d', 'D'],
        default: false
      },
      minSize: {
        default: 25000000,
        alias: ['min-size', 'minsize']
      }
    }
  }
);

async function main() {
  const log = new Logger(cli.flags.debug);
  const [rawInput, rawOutput] = cli.input;
  const cwd = process.cwd();
  const input = rawInput && path.resolve(cwd, rawInput);
  const output = rawOutput && input && path.resolve(cwd, rawOutput);

  if (!input || !output) {
    if (!input) {
      log.error('Missing input directory');
    }

    if (!output) {
      log.error('Missing output directory');
    }

    cli.showHelp();
    process.exit(1);
  }

  const inputExists = await pathExists(input);
  const outputExists = await pathExists(output);
  if (!inputExists) {
    log.throw(`Input directory does not exist`);
  }

  if (!outputExists) {
    await fs.promises.mkdir(output, { recursive: true });
  }

  const apollo = new Apollo({
    input,
    output,
    logger: log,
    debug: cli.flags.debug,
    minSize: cli.flags.minSize
  });

  await apollo.run();
}

main();
