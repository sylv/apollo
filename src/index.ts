import path from 'path';
import fs from 'fs';
import rrdir from 'rrdir';
import { ignoreFilesIncluding, mediaFileExtensions, stripFileNameRegex, supportingFileExtensions } from './constants';
import { stripNameTags } from './helpers/stripNameTags';
import { Logger } from './logger';
import { NameParser } from './parser';
import { FileType } from './parser/types';
import { ApolloOptions, File, GroupedFile, ParsedFile } from './types';

export class Apollo {
  private log = this.options.logger;
  private directories: string[] = [];

  constructor(readonly options: ApolloOptions & { input: string; output: string; logger: Logger }) {}

  /**
   * Run apollo using the options provided
   */
  async run() {
    const inputDisk = this.options.input.split(path.sep).filter(f => f)[0];
    const outputDisk = this.options.output.split(path.sep).filter(f => f)[0];
    if (inputDisk !== outputDisk) {
      // todo: handle this better - copying seems annoying and unreliable though.
      return this.log.throw('Input and output must be on the same disk.');
    }

    const files = await this.getFileIndex();
    const groupedFiles = this.groupFiles(files.parsedMediaFiles, files.parsedSupportingFiles);
    let checkedCount = 0;
    let newCount = 0;

    for (let group of groupedFiles) {
      const location = this.getFileLocation(group);
      const existingDir = this.directories.find(name => name.toLowerCase() === location.dir.toLowerCase());
      const locationDir = existingDir || location.dir;
      if (!existingDir) {
        await fs.promises.mkdir(locationDir, { recursive: true });
        this.directories.push(locationDir);
      } else {
        this.log.debug(`Skipping creation of directory "${locationDir}" as another one with the same case-insensitive name.`);
      }

      for (let file of group.files) {
        const ext = path.extname(file.path);
        const fileName = location.name + ext;
        const newFilePath = path.join(locationDir, fileName);

        this.log.debug(`Linking "${file.path}" -> "${newFilePath}"`);

        try {
          checkedCount += 1;
          await fs.promises.symlink(file.path, newFilePath);
          newCount += 1;
        } catch (e) {
          if (e.code === 'EPERM') {
            this.log.throw(
              `Permission error linking "${file.path}" -> "${newFilePath}". If you are on windows, make sure you are running Apollo as administrator.`
            );
          } else if (e.code === 'EEXIST') {
            this.log.debug(`File "${newFilePath}" already exists. Skipping symlink`);
            continue;
          }

          throw e;
        }
      }
    }

    this.log.info(`Checked ${checkedCount.toLocaleString()} files and moved  ${newCount.toLocaleString()} new files`);
  }

  /**
   * Get where a file should be saved to.
   */
  private getFileLocation(file: GroupedFile) {
    switch (file.parsed.type) {
      case FileType.EPISODE:
        const seasonName = file.parsed.seasonNumber ? `Season ${file.parsed.seasonNumber}` : 'Unknown Season';
        const episodeName = file.parsed.episodeName ? ` - ${file.parsed.episodeName}` : '';
        const seasonNumber = file.parsed.seasonNumber
          ? ` - s${file.parsed.seasonNumber.toString().padStart(2, '0')}e${file.parsed.episodeNumber?.toString().padStart(2, '0')}`
          : '';

        return {
          // "TV Shows/My Show/Season 1"
          dir: path.join(this.options.output, 'TV Shows', file.parsed.title as string, seasonName),
          // "My Show - 1x3" or "My Show - 1x3 - My Episode"
          name: `${file.parsed.title}${seasonNumber}${episodeName}`
        };
      case FileType.MOVIE:
        const name = `${file.parsed.title}` + (file.parsed.year ? ` (${file.parsed.year})` : '');

        return {
          // "Movies/My Movie" or "Movies/My Movie (2009)"
          dir: path.join(this.options.output, 'Movies', name),
          name
        };
    }
  }

  /**
   * Group "media" (video) files and "supporting" (subtitles, etc) files together.
   * @param mediaFiles
   * @param allSupportingFiles
   */
  private groupFiles(mediaFiles: ParsedFile[], allSupportingFiles: File[]) {
    const grouped: GroupedFile[] = [];
    for (const file of mediaFiles) {
      const supporting: File[] = [];

      switch (file.parsed.type) {
        case FileType.MOVIE:
          // for movies we're a little more lenient because there are much higher chances the only
          // supporting files in the directory are for the movie
          // basically we look for any supporting file starting with the movie title in the same directory
          const strippedTitle = this.stripStringForComparison(file.parsed.title as string);
          const startsWith = file.main.strippedParentDirectory.startsWith(strippedTitle);

          const check = (sFile: File): Boolean => {
            if (sFile.strippedName === file.main.strippedName) {
              return true;
            }

            // if the movie is in a directory for itself, we check subdirectories (e.g, "subttiles" is checked)
            // if the movie is not in a directory for itself (e.g, just sitting by itself in the input dir)
            // require the supporting file to be in the exact same dir as the main file
            const checkDir = startsWith ? sFile.directoryPath.startsWith(file.main.directoryPath) : sFile.directory === file.main.directory;
            if (checkDir && sFile.strippedName.startsWith(strippedTitle)) {
              return true;
            }

            return false;
          };

          const similarFiles = allSupportingFiles.filter(check);
          supporting.push(...similarFiles);

          break;
        default:
          const similar = allSupportingFiles.filter(s => s.strippedName === file.main.strippedName);
          supporting.push(...similar);
      }

      grouped.push({
        ...file,
        files: supporting.concat(file.main)
      });
    }

    return grouped;
  }

  /**
   * Get a list of all valid media and supporting files in this.options.input.
   */
  private async getFileIndex() {
    const parsedMediaFiles: ParsedFile[] = [];
    const parsedSupportingFiles: File[] = [];
    let ignoredCount = 0;

    for await (const file of rrdir.stream(this.options.input, { strict: true, followSymlinks: true, stats: true })) {
      const lowerPath = file.path.toLowerCase();
      const ignoreFile = ignoreFilesIncluding.some(part => lowerPath.includes(part));
      const isMediaFile = mediaFileExtensions.some(ext => file.path.endsWith(`.${ext}`));
      const isSupportingFile = supportingFileExtensions.some(ext => file.path.endsWith(`.${ext}`));
      const supportedType = isMediaFile || isSupportingFile;

      if (file.err || !file.stats || ignoreFile || !supportedType) {
        if (ignoreFile) {
          ignoredCount += 1;
          this.log.debug(`Ignoring "${file.path}" because it contains undesirable keywords`);
        } else if (!supportedType) {
          ignoredCount += 1;
          this.log.debug(`Ignoring "${file.path}" because it is not a supported file type`);
        } else if (file.err || !file.stats) {
          this.log.warn(`Ignoring "${file.path}" due to an error when reading the file`);
        }

        continue;
      }

      if (isMediaFile && file.stats.size < this.options.minSize) {
        ignoredCount += 1;
        this.log.debug(`Ignoring "${file.path}" as it is a media file smaller than ${this.options.minSize.toLocaleString()} bytes`);
        continue;
      }

      if (isSupportingFile) {
        const parsedSupportingFile = this.parseFilePath(file);
        parsedSupportingFiles.push(parsedSupportingFile);
      } else {
        const parsedMediaFile = this.parseMediaFile(file);
        if (!parsedMediaFile.parsed.title) {
          this.log.error(`Ignoring "${file.path}" because we failed to extract a title`);
          continue;
        }

        parsedMediaFiles.push(parsedMediaFile);
      }
    }

    if (ignoredCount !== 0 && !this.options.debug) {
      this.log.warn(`Ignored ${ignoredCount.toLocaleString()} undesirable files. Run with --debug for a full list.`);
    }

    return {
      parsedMediaFiles,
      parsedSupportingFiles
    };
  }

  /**
   * Parse a file path into structured data  using the name parser.
   * @param filePath The complete to the file, "D:\\My.Show.S05.COMPLETE.720p.WEB.x264-STRiFE[TGx]\\my.show.s05e12.my.episode.name720p.web.x264-strife.mkv"
   */
  private parseMediaFile(file: rrdir.Entry): ParsedFile {
    const main = this.parseFilePath(file);
    const parser = new NameParser();
    const parsed = parser.parse(main.fileName);

    return {
      main,
      parsed
    };
  }

  /**
   * Split up a file path into usable components
   * @param filePath The path to the file, "D:\\My.Show.S05.COMPLETE.720p.WEB.x264-STRiFE[TGx]\\my.show.s05e12.my.episode.name720p.web.x264-strife.mkv"
   */
  private parseFilePath(file: rrdir.Entry): File {
    const fileName = path.basename(file.path);
    const fileDirectory = path.dirname(file.path);
    const strippedFileName = this.stripStringForComparison(fileName);
    const parentDirectory = path.basename(fileDirectory);
    const strippedParentDirectory = this.stripStringForComparison(parentDirectory);

    return {
      ...file,
      fileName,
      directoryPath: fileDirectory,
      strippedName: strippedFileName,
      strippedParentDirectory
    };
  }

  /**
   * Strip undesirable parts of a string to make it better for direct comparison
   * @param input  The input, e.g "I.Am.Mother.2019.1080p.WEBRip.x264-[YTS.LT].mp4"
   * @returns "i am mother 2019 1080p webrip x264"
   */
  private stripStringForComparison(input: string) {
    return (
      stripNameTags(input.toLowerCase())
        // replace dots and underscores with spaces. we won't lose anything
        .replace(/\.|_/g, ' ')
        // replace things that might get in the way of direct comparisons
        .replace(stripFileNameRegex, ' ')
        // remove resolution, because sometimes "My.Movie.BluRay.720p.x264.srt" is intended for "My.Movie.2009.BluRay.1080p.x264"
        .replace(/[0-9]{3,4}p/i, ' ')
        // remove double spaces that we may have created
        .replace(/[ ]{2,}/g, ' ')
        .trim()
    );
  }
}
