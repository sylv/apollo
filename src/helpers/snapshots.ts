import Database from "better-sqlite3";
import { unlink } from "fs/promises";
import { Generated, Kysely, Selectable, SqliteAdapter, SqliteDialect } from "kysely";
import { homedir } from "os";
import { join } from "path";
import { moveFile } from "./move-file";

// .returning() with kysely will silently fail and return an empty object because it doesnt think
// sqlite supports returning *, but modern versions do, so this is a hacky workaround to get kysely to play nice.
Object.defineProperty(SqliteAdapter.prototype, "supportsReturning", {
  get: () => true,
});

interface EntryTable {
  id: Generated<number>;
  fromPath: string;
  toPath: string;
  jobId: number;
  symlink: 1 | 0;
  restored: 1 | 0;
}

export interface JobTable {
  id: Generated<number>;
  inputDirectory: string;
  outputDirectory: string;
  createdAt: number;
}

export interface Entry {
  id: number;
  fromPath: string;
  toPath: string;
  symlink: boolean;
  jobId: number;
}

export const snapshotDbPath = join(homedir(), ".config/apollo/snapshots.db");
const database = new Database(snapshotDbPath, { fileMustExist: false });
database.exec(`
  CREATE TABLE IF NOT EXISTS jobs (
    id INTEGER PRIMARY KEY,
    inputDirectory TEXT NOT NULL,
    outputDirectory TEXT NOT NULL,
    createdAt INTEGER NOT NULL
  );

  CREATE TABLE IF NOT EXISTS entries (
    id INTEGER PRIMARY KEY,
    fromPath TEXT NOT NULL,
    toPath TEXT NOT NULL,
    symlink INTEGER NOT NULL,
    restored INTEGER NOT NULL,
    jobId INTEGER NOT NULL REFERENCES jobs(id)
  );

  CREATE INDEX IF NOT EXISTS entries_jobId ON entries (jobId);
`);

const db = new Kysely<{ jobs: JobTable; entries: EntryTable }>({
  dialect: new SqliteDialect({
    database: database,
  }),
});

/**
 * Create a new job that can have entries appended to it.
 * @param inputDirectory The input directory where the files will be discovered
 * @param outputDirectory The output directory where the files will be moved to
 * @returns A new job to append entries to
 */
export async function createSnapshotJob(inputDirectory: string, outputDirectory: string) {
  const job = await db
    .insertInto("jobs")
    .values({
      inputDirectory: inputDirectory,
      outputDirectory: outputDirectory,
      createdAt: Date.now(),
    })
    .returningAll()
    .executeTakeFirstOrThrow();

  return job;
}

/**
 * Add an entry to a rename job.
 * This will allow the change to be rolled back in the future.
 * @param job The job to add the entry to.
 * @param from The location the file was moved from
 * @param to The location the file was moved to
 * @param isSymlink Whether the entry moved the file or symlinked it. Changes rollback behaviour.
 */
export async function addSnapshotEntryToJob(job: Selectable<JobTable>, from: string, to: string, isSymlink: boolean) {
  await db
    .insertInto("entries")
    .values({
      fromPath: from,
      toPath: to,
      jobId: job.id,
      symlink: isSymlink ? 1 : 0,
      restored: 0,
    })
    .execute();
}

/**
 * Get entries for the last job in the given input directory.
 * @see {@link rollback}
 */
export async function readLastJobEntries(inputDirectory: string): Promise<Entry[]> {
  // this was beautiful and used streaming and an asyncgenerator and was all cool and poggers
  // but then it turns out streaming doesnt work with sqlite :(
  let qb = db.selectFrom("jobs").orderBy("createdAt", "desc").selectAll();
  if (inputDirectory) {
    qb = qb.where("inputDirectory", "=", inputDirectory);
  }

  const latestJob = await qb.executeTakeFirst();
  if (!latestJob) return [];
  const results = await db
    .selectFrom("entries")
    .where("jobId", "=", latestJob.id)
    .where("restored", "!=", 1)
    .select(["id", "fromPath", "toPath", "symlink", "jobId"])
    .execute();

  return results.map((item) => ({
    id: item.id,
    fromPath: item.fromPath,
    toPath: item.toPath,
    symlink: item.symlink === 1,
    jobId: item.jobId,
  }));
}

/**
 * Roll back changes to the given input directory.
 * @returns the number of items rolled back, "0" means no entries were found.
 */
export const rollback = async (inputDirectory: string) => {
  const entries = await readLastJobEntries(inputDirectory);
  let restored = 0;
  for (const entry of entries) {
    if (entry.symlink) {
      await unlink(entry.toPath);
    } else {
      await moveFile({
        from: entry.toPath,
        to: entry.fromPath,
        symlink: false,
        dryRun: false,
        job: null,
      });
    }

    restored++;
    await db.updateTable("entries").where("id", "=", entry.id).set({ restored: 1 }).execute();
  }

  return restored;
};
