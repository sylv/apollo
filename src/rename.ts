import { stat } from "fs/promises";
import { basename, dirname, join } from "path";
import { ApolloParser, ApolloParserOptions } from "./classes/apollo-parser";
import { SUBTITLE_EXTENSIONS } from "./constants";
import { deleteEmptyDirs } from "./helpers/delete-empty-dirs";
import { moveFile } from "./helpers/move-file";
import { recursiveReaddir } from "./helpers/recursive-readdir";
import { replacePlaceholders } from "./helpers/replace-placeholders";
import { createSnapshotJob } from "./helpers/snapshots";
import { log } from "./log";
import { ApolloOutput, FileType, TitleType } from "./types";

export enum ApolloMode {
  Move,
  Symlink,
}

export enum ApolloStrategy {
  FileNames,
  Directory,
  Hybrid,
}

export interface RenameOptions extends ApolloParserOptions {
  inputDirectory: string;
  outputDirectory: string;
  strategy: ApolloStrategy;
  mode: ApolloMode;
  dryRun?: boolean;
  minSize: number | null;
  useSnapshots?: boolean;
  deleteEmptyDirs?: boolean;
  formats?: {
    episode?: string;
    movie?: string;
  };
}

export const defaultSnapshotFile = ".apollo-snapshot.json";
export const defaultEpisodeFormat = `TV Shows/{name} ({startYear}) [imdbId={imdbId}]/Season {seasonNumber}/{name} - S{seasonNumber}E{episodeNumber} - {episodeName}.{extension}`;
export const defaultMovieFormat = `Movies/{name} ({startYear}) [imdbId={imdbId}]/{name} ({startYear}).{extension}`;

function formatFilePath(options: RenameOptions, parsed: ApolloOutput) {
  if (parsed.fileType === FileType.Subtitle && parsed.languages?.length === 1) {
    // change extensions from ".srt" to ".en.srt" etc when we extract a single language
    // no need to convert to a code - languages property already attempts to do that from names like "french"
    parsed.extension = `.${parsed.languages[0]}${parsed.extension}`.toLowerCase();
  }

  if (parsed.titleType === TitleType.EPISODE) {
    const episodeFormat = options.formats?.episode ?? defaultEpisodeFormat;
    return replacePlaceholders(episodeFormat, parsed);
  } else if (parsed.titleType === TitleType.MOVIE) {
    const movieFormat = options.formats?.movie ?? defaultMovieFormat;
    return replacePlaceholders(movieFormat, parsed);
  } else {
    throw new Error(`Unknown title type "${parsed.titleType}"`);
  }
}

const STANDARD_SEASON_EPISODE_REGEX = /S[0-9]{1,4}E[0-9]{1,4}/;

function getFilePath(options: RenameOptions, filePath: string, parsed: ApolloOutput) {
  const formatted = formatFilePath(options, parsed);
  const originalFilename = basename(filePath);
  const useStrategy =
    options.strategy === ApolloStrategy.Hybrid
      ? STANDARD_SEASON_EPISODE_REGEX.test(originalFilename)
        ? ApolloStrategy.Directory
        : ApolloStrategy.FileNames
      : options.strategy;

  switch (useStrategy) {
    case ApolloStrategy.FileNames:
      return join(options.outputDirectory, formatted);
    case ApolloStrategy.Directory:
      const fullFormatted = join(options.outputDirectory, formatted);
      const withoutFilename = dirname(fullFormatted);
      return join(withoutFilename, originalFilename);
    default:
      throw new Error(`Unknown strategy "${useStrategy}"`);
  }
}

export async function renameFiles(options: RenameOptions) {
  const job = options.useSnapshots ? await createSnapshotJob(options.inputDirectory, options.outputDirectory) : null;
  for await (const filePath of recursiveReaddir(options.inputDirectory)) {
    const shortPath = filePath.slice(options.inputDirectory.length + 1);
    const isSubtitle = SUBTITLE_EXTENSIONS.some((ext) => filePath.endsWith(ext));
    if (!isSubtitle && options.minSize) {
      const meta = await stat(filePath);
      if (meta.size < options.minSize) {
        console.debug(`Skipping "${shortPath}" because it's too small`);
        continue;
      }
    }

    const strippedPath = filePath.slice(options.inputDirectory.length + 1);
    const parser = new ApolloParser(options);
    const parsed = await parser.parse(strippedPath, filePath);
    const isValidMovie = parsed?.titleType === TitleType.MOVIE;
    const isValidEpisode =
      parsed?.titleType === TitleType.EPISODE && parsed.seasonNumber !== undefined && parsed.episodeNumber?.length !== 0;

    if (!parsed?.name || (!isValidMovie && !isValidEpisode)) {
      log.warn(`Skipping "${shortPath}" as data could not be extracted`);
      continue;
    }

    const outputPath = getFilePath(options, filePath, parsed);
    await moveFile({
      dryRun: options.dryRun ?? false,
      from: filePath,
      to: outputPath,
      symlink: options.mode === ApolloMode.Symlink,
      job: job,
    });
  }

  if (options.deleteEmptyDirs) {
    log.info(`Deleting empty directories in "${options.inputDirectory}"`);
    await deleteEmptyDirs(options.inputDirectory);
  }
}
