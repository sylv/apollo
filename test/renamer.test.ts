jest.mock("fs/promises");
jest.mock("fs");

import { existsSync, readFileSync } from "fs";
import { DirectoryJSON, vol } from "memfs";
import { join } from "path";
import { rollback } from "../src/helpers/snapshots";
import { ApolloMode, ApolloStrategy, renameFiles, RenameOptions } from "../src/rename";

// in testing "Downloads-2" will break as it can't be removed because it might be important info,
// so the cli should slice that from the path.
const MOCK_DIRECTORY = "/mnt/Downloads-2/";
const MOCK_FILES = [
  "Deadpool 2016 1080p BluRay x264 DTS-JYK/Subs/French.srt",
  "The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E01.1080p.BluRay.x264-SHORTBREHD.mkv",
  "The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E02.1080p.BluRay.x264-SHORTBREHD.mkv",
  "The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E03.1080p.BluRay.x264-SHORTBREHD.mkv",
  "The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E04.1080p.BluRay.x264-SHORTBREHD.mkv",
  "The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E05.1080p.BluRay.x264-SHORTBREHD.mkv",
  "The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E06.1080p.BluRay.x264-SHORTBREHD.mkv",
  "The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E07.1080p.BluRay.x264-SHORTBREHD.mkv",
  "www.Torrenting.org       The Mandalorian S01E01 INTERNAL 1080p WEB H264-DEFLATE/The.Mandalorian.S01E01.INTERNAL.1080p.WEB.H264-DEFLATE.mkv",
  {
    name: "www.Torrenting.org       The Mandalorian S01E01 INTERNAL 1080p WEB H264-DEFLATE/The.Mandalorian.S01E01.INTERNAL.1080p.WEB.H264-DEFLATE.srt",
    content: readFileSync(join(__dirname, "./fixtures/lalaland.srt"), "utf8"),
  },
].map((file) => {
  if (typeof file === "string") {
    return {
      path: `${MOCK_DIRECTORY}${file}`,
      content: "",
    };
  } else {
    return {
      path: `${MOCK_DIRECTORY}${file.name}`,
      content: file.content,
    };
  }
});

beforeEach(() => {
  const config: DirectoryJSON = {};
  for (const file of MOCK_FILES) config[file.path] = file.content;
  vol.fromJSON(config);
});

afterEach(() => {
  vol.reset();
});

export const baseRenameOptions: RenameOptions = {
  inputDirectory: "/mnt/Downloads-2",
  outputDirectory: "/mnt/media",
  providers: ["local"],
  strategy: ApolloStrategy.FileNames,
  mode: ApolloMode.Move,
  minSize: 0,
  detectSubtitleLanguage: true,
  useSnapshots: false,
};

it("should move files as expected", async () => {
  await renameFiles({
    ...baseRenameOptions,
    deleteEmptyDirs: true,
  });

  expect(Object.keys(vol.toJSON())).toMatchSnapshot();
});

it("should not do anything with --dry-run enabled", async () => {
  await renameFiles({
    ...baseRenameOptions,
    dryRun: true,
  });

  expect(Object.keys(vol.toJSON())).toMatchSnapshot();
  expect(existsSync(`/mnt/media`)).toBe(false);
});

it("should allow reverting from snapshots", async () => {
  await renameFiles({
    ...baseRenameOptions,
    useSnapshots: true,
    deleteEmptyDirs: true,
  });

  expect(Object.keys(vol.toJSON())).toMatchSnapshot();
  await rollback(baseRenameOptions.inputDirectory);
  expect(Object.keys(vol.toJSON())).toMatchSnapshot();
});
