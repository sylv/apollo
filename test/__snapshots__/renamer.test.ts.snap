// Vitest Snapshot v1

exports[`should allow reverting from snapshots 1`] = `
Array [
  "/mnt/Downloads-2",
  "/mnt/media/Movies/Deadpool (2016) [imdbId=tt1431045]/Deadpool (2016).fr.srt",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E01 - Wolferton Splash.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E02 - Hyde Park Corner.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E03 - Windsor.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E04 - Act of God.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E05 - Smoke and Mirrors.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E06 - Gelignite.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E07 - Scientia Potentia Est.mkv",
  "/mnt/media/TV Shows/The Mandalorian (2019) [imdbId=tt8111088]/Season 1/The Mandalorian - S01E01 - Chapter 1 The Mandalorian.mkv",
  "/mnt/media/TV Shows/The Mandalorian (2019) [imdbId=tt8111088]/Season 1/The Mandalorian - S01E01 - Chapter 1 The Mandalorian.en.srt",
]
`;

exports[`should allow reverting from snapshots 2`] = `
Array [
  "/mnt/Downloads-2/Deadpool 2016 1080p BluRay x264 DTS-JYK/Subs/French.srt",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E01.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E02.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E03.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E04.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E05.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E06.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E07.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/www.Torrenting.org       The Mandalorian S01E01 INTERNAL 1080p WEB H264-DEFLATE/The.Mandalorian.S01E01.INTERNAL.1080p.WEB.H264-DEFLATE.mkv",
  "/mnt/Downloads-2/www.Torrenting.org       The Mandalorian S01E01 INTERNAL 1080p WEB H264-DEFLATE/The.Mandalorian.S01E01.INTERNAL.1080p.WEB.H264-DEFLATE.srt",
  "/mnt/media/Movies/Deadpool (2016) [imdbId=tt1431045]",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1",
  "/mnt/media/TV Shows/The Mandalorian (2019) [imdbId=tt8111088]/Season 1",
]
`;

exports[`should move files as expected 1`] = `
Array [
  "/mnt/Downloads-2",
  "/mnt/media/Movies/Deadpool (2016) [imdbId=tt1431045]/Deadpool (2016).fr.srt",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E01 - Wolferton Splash.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E02 - Hyde Park Corner.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E03 - Windsor.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E04 - Act of God.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E05 - Smoke and Mirrors.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E06 - Gelignite.mkv",
  "/mnt/media/TV Shows/The Crown (2016) [imdbId=tt4786824]/Season 1/The Crown - S01E07 - Scientia Potentia Est.mkv",
  "/mnt/media/TV Shows/The Mandalorian (2019) [imdbId=tt8111088]/Season 1/The Mandalorian - S01E01 - Chapter 1 The Mandalorian.mkv",
  "/mnt/media/TV Shows/The Mandalorian (2019) [imdbId=tt8111088]/Season 1/The Mandalorian - S01E01 - Chapter 1 The Mandalorian.en.srt",
]
`;

exports[`should not do anything with --dry-run enabled 1`] = `
Array [
  "/mnt/Downloads-2/Deadpool 2016 1080p BluRay x264 DTS-JYK/Subs/French.srt",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E01.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E02.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E03.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E04.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E05.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E06.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/The.Crown.S01.1080p.BluRay.x264-SHORTBREHD[rartv]/The.Crown.S01E07.1080p.BluRay.x264-SHORTBREHD.mkv",
  "/mnt/Downloads-2/www.Torrenting.org       The Mandalorian S01E01 INTERNAL 1080p WEB H264-DEFLATE/The.Mandalorian.S01E01.INTERNAL.1080p.WEB.H264-DEFLATE.mkv",
  "/mnt/Downloads-2/www.Torrenting.org       The Mandalorian S01E01 INTERNAL 1080p WEB H264-DEFLATE/The.Mandalorian.S01E01.INTERNAL.1080p.WEB.H264-DEFLATE.srt",
]
`;
